### PR TITLE
Major update to Natacha's pull request.

### PR DESCRIPTION
**Pull request by prioux**
The tests are failing but I will look at the tomorrow. In the meantime:

1- pull the code on your side
2- add the necessary info in oidc.yml.erb to configure a globus and keycloak server
3- verify that all login options work normally, including globus.

### CODE REVIEW

**Comment by prioux** [comment]
Okay I know why the tests are failing; in a test environment, not all validations are run, and we load the  OIDC config file during validations. So we got a `nil` instead of a `[]` . Anyway, easy fix. Will commit in just a few minutes.

**Comment by natacha-beck** [comment]
Tested on my side with my local KC installation and with Globus. Both work as expected.

File: BrainPortal/lib/oidc_config.rb [hunk 1]
```diff
@@ -20,84 +20,117 @@
...
+
+  REQUIRED_KEYS = %w(
+       authorize_uri token_uri logout_uri
+       client_id client_secret
+       scope
+       identity_provider_key identity_provider_display_name_key identity_preferred_username_key
+  )
+
+  # This populate a static array of configurations in the class, as loaded
+  # from a YAML.erb file. See RAILS_ROOT/config/oidc.yaml.erb for an example.
```
Comment (natacha-beck) [inline BrainPortal/lib/oidc_config.rb:50]: We should decide between `yaml` or `yml`

File: BrainPortal/config/oidc.yml.erb [hunk 2]
```diff
@@ -1,50 +1,73 @@
+
+#
+# CBRAIN Identity Provider Configuration File "oidc.yml.erb"
```
Comment (natacha-beck) [inline BrainPortal/config/oidc.yml.erb:3]: Maybe `yaml.erb` file.

File: BrainPortal/config/oidc.yml.erb [hunk 3]
```diff
@@ -1,50 +1,73 @@
...
+  client_id:                          "put_the_client_id_here"
+  client_secret:                      <%= ENV.delete("KEYCLOAK_CLIENT_SECRET") %>
+  # URLs to connect to the ID provider. The authorize URL provided the login button (and will be provided by
+  # special query params at run time). The token URL is internal during negotiation. The logout
+  # URL is fixed and just logs the user out of the provider's session.
+  # For Keycloak and other OpenID providers, you must adjust REALM_NAME !
+  authorize_uri:                      "http://localhost:16381/realms/REALM_NAME/protocol/openid-connect/auth"
+  token_uri:                          "http://localhost:16381/realms/REALM_NAME/protocol/openid-connect/token"
+  logout_uri:                         "http://localhost:16381/realms/REALM_NAME/protocol/openid-connect/logout"
+  # Scope used when called the OpenID provider when autheticate
```
Comment (natacha-beck) [inline BrainPortal/config/oidc.yml.erb:46]: Typo in `autheticate`

File: BrainPortal/config/oidc.yml.erb [hunk 4]
```diff
@@ -1,50 +1,73 @@
...
-  client_id:                          <%= ENV.delete("globus_client_id") %>
-  client_secret:                      <%= ENV.delete("globus_secret_key") %>
-  # Will be used to extract information from the returned identity
-  identity_provider_key:              'identity_provider'
-  identity_provider_display_name_key: 'identity_provider_display_name'
-  identity_preferred_username_key:    'preferred_username'
-  # help_label and help_uri are used for documentation and it's optional
-  help_label:                        'Globus'
-  help_uri:                          'https://support.globus.org/hc/en-us'
+  scope:                              "openid email profile"
```
Comment (natacha-beck) [inline BrainPortal/config/oidc.yml.erb:66]: Are you sure about the scope it is not the same as the initial one.

File: BrainPortal/lib/globus_helpers.rb [hunk 5]
```diff
@@ -229,24 +194,29 @@ def find_users_with_specific_identity(identity, oidc)
...
+  def unlink_oidc_identity(oidc, user)
+    oidc.zap_linked_oidc_info(user) # prov_id, prov_name, username set to NIL
     user.addlog("Unlinked #{oidc.name} identity")
   end
 
-
+  # Returns a hash table with keys being the names of the OidcConfigs
+  # and values being the login URL that includes the redirect callback URL.
+  # This is used by the interface to generate login buttons.
   def generate_oidc_login_uri(oidc_providers, redirect_url) #:nodoc:
```
Comment (natacha-beck) [inline BrainPortal/lib/globus_helpers.rb:216]: Maybe remove the :nodoc

File: BrainPortal/lib/oidc_config.rb [hunk 6]
```diff
@@ -20,84 +20,117 @@
...
 
     loaded_yaml = YAML.load(ERB.new(File.read(path)).result).with_indifferent_access
 
-    needed_keys = %w[enabled authorize_uri token_uri logout_uri scope client_secret client_id
-                     identity_provider_key identity_provider_display_name_key identity_preferred_username_key]
-
-    errors = []
     loaded_yaml.each do |name, config|
 
+      next if ! config[:enabled] # skip entirely enything not enabled.
```
Comment (natacha-beck) [inline BrainPortal/lib/oidc_config.rb:65]: Typo in `enything`

File: BrainPortal/lib/oidc_config.rb [hunk 7]
```diff
@@ -107,48 +140,104 @@ def self.find_by_state(state)
...
-    user.meta[self.provider_id_key()] = provider_id
+  # -------------------------------------------------------
+  # Identity structure KEY methods.
+  # -------------------------------------------------------
+  # These methods return the key names (as symbols)
+  # that are used to extract out of a OpenID identity
+  # structure these three values:
+  # provider ID, provider name, and a username.
+
+  def provider_id_key #:nodoc:
```
Comment (natacha-beck) [inline BrainPortal/lib/oidc_config.rb:166]: Remove :nodoc ?

File: BrainPortal/lib/portal_system_checks.rb [hunk 8]
```diff
@@ -93,6 +93,24 @@ def self.a020_check_database_sanity #:nodoc:
...
 
+  def self.a030_ensure_we_can_load_oidc_config
+
+      #----------------------------------------------------------------------------
+      puts "C> Loading OIDC configuration, if any..."
+      #----------------------------------------------------------------------------
+
+      begin
+        OidcConfig.load_from_file
+        puts "C> \t- Found ODIC Config(s): #{OidcConfig.all_names.join(", ")}"
```
Comment (natacha-beck) [inline BrainPortal/lib/portal_system_checks.rb:104]: Typo in `ODIC`

**Comment by prioux** [comment]
Comments applied to #8

**Pull request closed without merging**